synapse/handlers/deactivate_account.py

```diff
@@ -136,10 +136,19 @@ async def deactivate_account(
         # Mark the user as erased, if they asked for that
         if erase_data:
             user = UserID.from_string(user_id)
-            # Remove avatar URL from this user
-            await self._profile_handler.set_avatar_url(user, requester, "", by_admin)
-            # Remove displayname from this user
-            await self._profile_handler.set_displayname(user, requester, "", by_admin)
+            try:
+                # Remove avatar URL from this user
+                await self._profile_handler.set_avatar_url(
+                    user, requester, "", by_admin
+                )
+                # Remove displayname from this user
+                await self._profile_handler.set_displayname(
+                    user, requester, "", by_admin
+                )
+            except StoreError as e:
+                # bypass only 404 if user does not have a profile
+                if e.code != 404:
+                    raise e
```

Comment:
So the reason a `StoreError` is raised here is due to `set_profile_avatar_url` performing an `update` on `profiles` rather than an `upsert`:

https://github.com/matrix-org/synapse/blob/4b965c862dc66c0da5d3240add70e9b5f0aa720b/synapse/storage/databases/main/profile.py#L83-L91

I think a more fundamental fix for this whole problem would actually just be to convert `set_profile_avatar_url` to use `self.db_pool.simple_upsert`, like so:

```python
        await self.db_pool.simple_upsert(
            table="profiles",
            keyvalues={"user_id": user_localpart},
            values={},
            insertion_values={"avatar_url": new_avatar_url},
            desc="set_profile_avatar_url",
        )
```

it'd probably be a good idea to do the same for `set_profile_displayname` as well, just to avoid the same issue in the future.

Reply:
Ok. I will make an update.